src/chain/chain.rs

```diff
@@ -513,6 +531,43 @@ impl Chain {
         self.our_prefix().is_neighbour(sec_info.prefix()) && self.is_new(sec_info)
     }
 
+    /// Returns the index of the public key in our_history that will be trusted by the target
+    /// Authority
+    fn proving_index(&self, target: &Authority<XorName>) -> u64 {
+        self.their_knowledge
+            .iter()
+            .find(|(prefix, _)| prefix.matches(&target.name()))
+            .map(|(_, index)| *index)
+            .unwrap_or(0)
+    }
+
+    /// Provide a SectionProofChain that proves the given signature to the section with a given
+    /// prefix
+    pub fn prove(
+        &self,
+        target: &Authority<XorName>,
+        signature: &BlsSignature,
+        data: &[u8],
+    ) -> SectionProofChain {
+        let first_index = self.proving_index(target);
+        let keys_len = self.state.our_history.len();
+        let last_index = self
```

Comment:
Do we need that all find thing because we are attaching the proof on the relay node as oppose to when each node sign the message. (Similar to other comment, if messages are created on consnesus event, they will have all the same proof.). So this can be a lot simpler I guess (i.e only need to look for the first_index, I guess the last index is the one we use to sign things).

Reply:
Well, it might not be strictly necessary, but it's just to make sure that the last public key actually correctly validates the signature.

This can be done another way, too. If we include the `PublicKeySet` in the `Partial` variant, it would also carry the information about which public key is the correct one. So if we use this information and store it in the `Signed` variant, too, we can pass it here when we're attaching the proof - and there will be no need to look for it.

Reply:
Related to https://github.com/maidsafe/routing/pull/1721#discussion_r311943840, since we do it when we create the signature we know it is the last one.